core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/WindowsShareCopy.java

```diff
@@ -0,0 +1,267 @@
+public class WindowsShareCopy extends Action {
+  private static final Logger LOG = LoggerFactory.getLogger(WindowsShareCopy.class);
+  private static final long MIN_BUFFER_SIZE = 4096;
+  private static final int MIN_NUM_THREADS = 1;
+  private WindowsShareCopyConfig config;
+
+  public WindowsShareCopy(WindowsShareCopyConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void run(ActionContext context) throws Exception {
+    StringBuilder sb = new StringBuilder("smb://");
+    sb.append(config.netBiosHostname);
+    sb.append("/");
+    sb.append(config.netBiosSharename);
+    sb.append("/");
+    if (config.sourceDirectory.startsWith("/")) {
+      if (config.sourceDirectory.length() > 1) {
+        sb.append(config.sourceDirectory.substring(1));
+      }
+    } else {
+      sb.append(config.sourceDirectory);
+    }
+    final String smbDirectory = sb.toString();
+
+    // Register the SMB File handler.
+    jcifs.Config.registerSmbURLHandler();
+    jcifs.Config.setProperty("jcifs.util.loglevel", "4");
+
+    // Authentication with NTLM and read the directory from the Windows Share.
+    final NtlmPasswordAuthentication auth = new NtlmPasswordAuthentication(config.netBiosDomainName,
+                                                                           config.netBiosUsername,
+                                                                           config.netBiosPassword);
+    SmbFile dir = new SmbFile(smbDirectory, auth);
+
+    // Create the HDFS directory if it doesn't exist.
+    final Path hdfsDir = new Path(config.destinationDirectory);
+    final FileSystem hdfs = hdfsDir.getFileSystem(new Configuration());
+    if (!hdfs.exists(hdfsDir)) {
+      hdfs.mkdirs(hdfsDir);
+    }
+
+    // Determines the buffer size to be used during writing.
+    Configuration conf = new Configuration();
+    conf.setLong("io.file.buffer.size", config.bufferSize);
+    String[] files = dir.list();
+
+    // Copies the files in a multithreaded way
+    CountDownLatch executorTerminateLatch = new CountDownLatch(1);
+    ExecutorService executorService = createExecutor(config.numThreads, executorTerminateLatch);
+    CompletionService<String> completionService = new ExecutorCompletionService<>(executorService);
+
+    try {
+      for (final String file : files) {
+        completionService.submit(new Callable<String>() {
+          @Override
+          public String call() throws Exception {
+            try {
+              if (smbDirectory.endsWith("/")) {
+                return copyFileToHDFS(hdfs, smbDirectory + file, hdfsDir, auth);
+              } else {
+                return copyFileToHDFS(hdfs, smbDirectory + "/" + file, hdfsDir, auth);
+              }
+            } catch (Exception e) {
+              LOG.warn("Exception while copying the file {}", file, e);
+              return null;
+            }
+          }
+        });
+      }
+
+      for (int i = 0; i < files.length; i++) {
+        try {
```

Comment:
Also the value of i is not used anywhere so don't see the need for loop. I guess you are trying to make sure that files.length job completed in which case keeping a running count till it reached files.length will be more suitable.